cibyl/features/__init__.py

```diff
@@ -0,0 +1,156 @@
+    return is_concrete_class and "features" in symbol.__module__
+
+
+class FeaturesLoader:
+    """Load all features available, either general or product specific through
+    some plugin."""
+    # __path__ is a list
+    features_locations = __path__
+
+    def __init__(self):
+        self.features_by_category = defaultdict(list)
+        self.all_features = {}
+        for location in self.features_locations:
+            for module_path in os.listdir(location):
+                module_match = MODULE_PATTERN.match(module_path)
+                if not module_match:
+                    continue
+                module_name = module_match.group(1)
+                # import the module directly from file, and add the _features
+                # suffix to the imported modules to help distinguish the
+                # classes defined there
+                # https://docs.python.org/3/library/importlib.html#importing-a-source-file-directly
+                spec = spec_from_file_location(module_name+"_features",
+                                               f"{location}/{module_path}")
+                module = module_from_spec(spec)
+                spec.loader.exec_module(module)
+                features = getmembers(module, predicate=is_feature_class)
+                for feature_name, feature in features:
+                    self.all_features[feature_name.lower()] = feature
+                    self.features_by_category[module_name].append(feature_name)
+
+    def get_string_all_features(self):
+        """Get a string representation listing all available features to use in
+        an exception message."""
+        msg = ""
+        for category, features_names in self.features_by_category.items():
+            msg += f"\n{Colors.blue(category.title())}:"
+            for feature_name in features_names:
+                feature = self.all_features[feature_name.lower()]
+                docstring = getattr(feature, "__doc__", "")
+                msg += f"\n{Colors.red(' * ')}{Colors.blue(feature_name)}"
+                if docstring:
+                    msg += f"{Colors.red(' - '+feature.__doc__)}"
+        return msg
+
+    def get_feature(self, name_feature):
+        """Get the function associated with the given feature name
+
+        :param feature_name: Name of the feature
+        :type feature_name: str
+
+        :returns: class that implements the given feature
+        :rtype: class
+        :raises: InvalidArgument
+        """
+        try:
+            feature_class = self.all_features[name_feature.lower()]
+        except KeyError as err:
+            msg = f"No feature {name_feature}. Choose one of the following "
+            msg += "features:\n"
+            msg += self.get_string_all_features()
+            raise InvalidArgument(msg) from err
+        return feature_class()
+
+
+class FeatureTemplate(ABC):
+    """Skeleton for a generic feature."""
+    method_to_query = ""
+
+    @abstractmethod
```

Comment:
I have never seen an abstract constructor before, they cannot be overridden. What do you want to achieve here?

Reply:
Actually the reason the __init__ was marked as abstract was kind of an ugly hack, in order for the isabstract function of inspect to work, the class needed at least one abstract method, with your suggestions this is not necessary